lib/analyzer/image-inspector.ts

```diff
@@ -12,6 +17,130 @@ async function detect(
   return JSON.parse(info.stdout)[0];
 }
 
+/**
+ * In the case that an `ImageType.Identifier` is detected we need to produce
+ * an image archive, either by saving the image if it's already loaded into
+ * the local docker daemon, or by pulling the image from a remote registry and
+ * saving it to the filesystem directly.
+ *
+ * Users may also provide us with a URL to an image in a Docker compatible
+ * remote registry.
+ *
+ * @param {string} targetImage - The image to test, this could be in one of
+ *    the following forms:
+ *      * [registry/]<repo>/<image>[:tag]
+ *      * <repo>/<image>[:tag]
+ *      * <image>[:tag]
+ *    In the case that a registry is not provided, the plugin will default
+ *    this to Docker Hub. If a tag is not provided this will default to
+ *    `latest`.
+ * @param {string} [username] - Optional username for private repo auth.
+ * @param {string} [password] - Optional password for private repo auth.
+ */
+async function getImageArchive(
+  targetImage: string,
+  username?: string,
+  password?: string,
+): Promise<ArchiveResult> {
+  const docker = new Docker(targetImage);
+  const destination = tmp.dirSync({ unsafeCleanup: true });
+  const saveLocation: string = path.join(destination.name, "image.tar");
+
+  try {
+    // TODO: Eventually need to provide support for specifying a user-defined
+    // temporary directory.
+    await docker.save(targetImage, saveLocation);
+    return {
+      path: saveLocation,
+      removeArchive: destination.removeCallback,
+    };
+  } catch {
+    debug(
+      `${targetImage} does not exist locally, proceeding to pull image from remote registry.`,
+    );
+  }
+
+  if (await Docker.binaryExists()) {
+    try {
+      if (username || password) {
+        debug(
+          "using local docker binary credentials. the credentials you provided will be ignored",
+        );
+      }
+
+      await docker.pullCli(targetImage);
+      await docker.save(targetImage, saveLocation);
+      return {
+        path: saveLocation,
+        removeArchive: destination.removeCallback,
+      };
+    } catch (err) {
+      debug(`couldn't pull ${targetImage} using docker binary: ${err}`);
+    }
+  }
+
+  destination.removeCallback();
+  const { hostname, imageName, tag } = await extractImageDetails(targetImage);
+
+  debug(
+    `Attempting to pull: registry: ${hostname}, image: ${imageName}, tag: ${tag}`,
+  );
+  const pullResult = await docker.pull(
+    hostname,
+    imageName,
+    tag,
+    username,
+    password,
+  );
+  return {
+    path: path.join(pullResult.stagingDir!.name, "image.tar"),
+    removeArchive: pullResult.stagingDir!.removeCallback,
```

Comment:
Can these be undefined? Is the pull library type lying to us and can we fix it in the library so we don't have to force a cast?

Reply:
The return type from the lib is `stagingDir: tmp.DirResult | null;`. I'm not sure if this can be changed, but myself or @shaimendel will check tomorrow.

Reply:
I find this issue less critical, wdyt?

Reply:
Agree